File: library/src/main/java/com/senierr/adapter/support/wrapper/BaseStateWrapper.java
package com.senierr.adapter.support.wrapper;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;

import com.senierr.adapter.internal.MultiTypeAdapter;
import com.senierr.adapter.internal.ViewHolderWrapper;
import com.senierr.adapter.support.bean.StateBean;

/**
 * 状态封装
 *
 * @author zhouchunjie
 * @date 2017/10/9
 */
public abstract class BaseStateWrapper extends ViewHolderWrapper<StateBean> {

    private @Nullable RecyclerView recyclerView;
    private @NonNull StateBean stateBean;

    public BaseStateWrapper() {
        stateBean = new StateBean();
    }

    @Override
    public int getSpanSize(StateBean item) {
        if (recyclerView == null) {
            return super.getSpanSize(item);
        }
        return Integer.MAX_VALUE;
    }

    @Override
    public void onAttachedToRecyclerView(RecyclerView recyclerView) {
        super.onAttachedToRecyclerView(recyclerView);
        this.recyclerView = recyclerView;
    }

    /**
     * 更新状态
     *
     * @param state
     */
    public void setState(int state) {
        stateBean.setState(state);
        MultiTypeAdapter multiTypeAdapter = getMultiTypeAdapter();
        if (multiTypeAdapter != null && recyclerView != null) {
            multiTypeAdapter.getDataList().clear();
            multiTypeAdapter.getDataList().add(stateBean);
            multiTypeAdapter.notifyDataSetChanged();
            RecyclerView.LayoutManager layoutManager = recyclerView.getLayoutManager();
            if (layoutManager != null) {
                if (layoutManager instanceof StaggeredGridLayoutManager) {
                    ((StaggeredGridLayoutManager) layoutManager).invalidateSpanAssignments();
                }
                layoutManager.scrollToPosition(0);
            }
        }
    }
}

File: multitype/src/main/java/com/senierr/adapter/support/wrapper/BaseStateWrapper.kt
package com.senierr.adapter.support.wrapper

import android.support.annotation.LayoutRes
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.StaggeredGridLayoutManager
import com.senierr.adapter.internal.ViewHolderWrapper
import com.senierr.adapter.support.bean.StateBean

/**
 * 状态封装
 *
 * @author zhouchunjie
 * @date 2017/10/9
 */
abstract class BaseStateWrapper(@LayoutRes private val layoutId: Int = -1) : ViewHolderWrapper<StateBean>(layoutId) {

    private var recyclerView: RecyclerView? = null
    private val stateBean = StateBean()

    override fun getSpanSize(item: StateBean): Int = Integer.MAX_VALUE

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)
        this.recyclerView = recyclerView
    }

    /**
     * 更新状态
     *
     * @param state
     */
    fun setState(state: Int) {
        stateBean.state = state
        recyclerView?.let {
            multiTypeAdapter.data.clear()
            multiTypeAdapter.data.add(stateBean)
            multiTypeAdapter.notifyDataSetChanged()
            val layoutManager = it.layoutManager
            if (layoutManager != null) {
                if (layoutManager is StaggeredGridLayoutManager) {
                    layoutManager.invalidateSpanAssignments()
                }
                layoutManager.scrollToPosition(0)
            }
        }
    }
}
